From 118dfff69bb158a38767d668e0fbf91f47dd38f6 Mon Sep 17 00:00:00 2001
From: Adam Stolcenburg <adam_stolcenburg@comcast.com>
Date: Fri, 8 May 2026 16:35:10 +0200
Subject: [PATCH] RDKEAPPRT-762 Add reading of cert scope and secret key

This change also hard code some model properties, to start
the certification process.
---
 .../starboard/rdk/arm/configuration_public.h  |  7 +-
 src/third_party/starboard/rdk/shared/BUILD.gn |  1 +
 .../rdk/shared/system/system_get_property.cc  |  6 +-
 ...stem_sign_with_certification_secret_key.cc | 79 +++++++++++++++++++
 4 files changed, 87 insertions(+), 6 deletions(-)

diff --git a/src/third_party/starboard/rdk/arm/configuration_public.h b/src/third_party/starboard/rdk/arm/configuration_public.h
index 98cee42..85fb7fc 100644
--- a/src/third_party/starboard/rdk/arm/configuration_public.h
+++ b/src/third_party/starboard/rdk/arm/configuration_public.h
@@ -302,9 +302,10 @@
 #endif
 
 #define SB_PLATFORM_FRIENDLY_NAME "RDK"
-#define SB_PLATFORM_OPERATOR_NAME "Unknown"
-#define SB_PLATFORM_MODEL_YEAR 2021
-#define SB_PLATFORM_MODEL_NAME "RDKReference"
+#define SB_PLATFORM_OPERATOR_NAME "RDKCommonPort"
+#define SB_PLATFORM_MANUFACTURER_NAME "RDKCommonPort"
+#define SB_PLATFORM_MODEL_YEAR 2026
+#define SB_PLATFORM_MODEL_NAME "BCM74166SFF"
 #define SB_PLATFORM_FIRMWARE_VERSION_STRING "1.0.0"
 
 #endif  // THIRD_PARTY_STARBOARD_RDK_ARM_CONFIGURATION_PUBLIC_H_
diff --git a/src/third_party/starboard/rdk/shared/BUILD.gn b/src/third_party/starboard/rdk/shared/BUILD.gn
index 86ace61..69460c0 100644
--- a/src/third_party/starboard/rdk/shared/BUILD.gn
+++ b/src/third_party/starboard/rdk/shared/BUILD.gn
@@ -440,6 +440,7 @@ static_library("starboard_platform") {
     "GLESv2",
     "essos",
     "dl",
+    "crypto",
   ]
 
   configs += [
diff --git a/src/third_party/starboard/rdk/shared/system/system_get_property.cc b/src/third_party/starboard/rdk/shared/system/system_get_property.cc
index e380c90..f847828 100644
--- a/src/third_party/starboard/rdk/shared/system/system_get_property.cc
+++ b/src/third_party/starboard/rdk/shared/system/system_get_property.cc
@@ -143,7 +143,7 @@ bool GetChipsetModelNumber(char* out_value, int value_length) {
     return CopyStringAndTestIfSuccess(out_value, value_length, chipset->c_str());
   }
 
-  return false;
+  return CopyStringAndTestIfSuccess(out_value, value_length, "74116");
 }
 
 bool GetFirmwareVersion(char* out_value, int value_length) {
@@ -155,13 +155,13 @@ bool GetFirmwareVersion(char* out_value, int value_length) {
     return CopyStringAndTestIfSuccess(out_value, value_length, firmware_version->c_str());
   }
 
-  return false;
+  return CopyStringAndTestIfSuccess(out_value, value_length, "0.0.1");
 }
 
 bool GetCertificationScope(char* out_value, int value_length) {
   const char *cert_scope_file_name = std::getenv("COBALT_CERT_SCOPE_FILE_NAME");
   if ( cert_scope_file_name == nullptr )
-    cert_scope_file_name = "/opt/drm/0681000006810001.bin";
+    cert_scope_file_name = "/opt/cert_scope";
 
   ::starboard::ScopedFile file(cert_scope_file_name, O_RDONLY);
   if ( !file.IsValid() ) {
diff --git a/src/third_party/starboard/rdk/shared/system/system_sign_with_certification_secret_key.cc b/src/third_party/starboard/rdk/shared/system/system_sign_with_certification_secret_key.cc
index 762e56f..b6ad3b0 100644
--- a/src/third_party/starboard/rdk/shared/system/system_sign_with_certification_secret_key.cc
+++ b/src/third_party/starboard/rdk/shared/system/system_sign_with_certification_secret_key.cc
@@ -36,10 +36,13 @@
 #include "starboard/system.h"
 #include "starboard/string.h"
 #include "starboard/memory.h"
+#include "starboard/common/file.h"
 
 #include "third_party/starboard/rdk/shared/hang_detector.h"
 #include "third_party/starboard/rdk/shared/log_override.h"
 
+#include "third_party/modp_b64/modp_b64.h"
+
 #if defined(HAS_WPECRYPTOGRAPHY)
 #include <cryptography/cryptography.h>
 #if defined(HAS_RFC_API)
@@ -47,6 +50,14 @@
 #endif
 #endif
 
+#include <openssl/hmac.h>
+
+size_t modp_b64_decode(
+    char* dest,
+    const char* src,
+    size_t len,
+    ModpDecodePolicy policy) __attribute__((weak));
+
 namespace {
 
 template<typename T>
@@ -57,12 +68,79 @@ struct RefDeleter {
 template<typename T>
 using ScopedRef = std::unique_ptr<T, RefDeleter<T>>;
 
+/* code imported from: https://github.com/youtube/cobalt/blob/main/starboard/shared/deviceauth/deviceauth_internal.cc */
+bool Base64Decode(const std::string& input, std::string* output) {
+  std::string temp;
+  temp.resize(modp_b64_decode_len(input.size()));
+
+  // does not null terminate result since result is binary data!
+  size_t input_size = input.size();
+  int output_size = modp_b64_decode(&(temp[0]), input.data(), input_size);
+  if (output_size == MODP_B64_ERROR)
+    return false;
+
+  temp.resize(output_size);
+  output->swap(temp);
+  return true;
+}
+
+bool SignWithCertificationSecretKey(const char* secret_key,
+                                    const uint8_t* message,
+                                    size_t message_size_in_bytes,
+                                    uint8_t* digest,
+                                    size_t digest_size_in_bytes) {
+  if (digest_size_in_bytes < 32) {
+    SB_LOG(ERROR) << "SbSystemSignWithCertificationSecretKey() must be called "
+                  << "with an output digest buffer size of at least 32 bytes.";
+    return false;
+  }
+  const size_t kBase64EncodedCertificationSecretLength = 1023;
+  char base_64_secret_property[kBase64EncodedCertificationSecretLength + 1] = {
+      0};
+  strncpy(base_64_secret_property, secret_key,
+          kBase64EncodedCertificationSecretLength);
+
+  std::string secret;
+  SB_CHECK(Base64Decode(base_64_secret_property, &secret));
+
+  uint8_t* result = HMAC(EVP_sha256(), secret.data(), secret.size(), message,
+                         message_size_in_bytes, digest, nullptr);
+
+  return result == digest;
+}
+
 }
 
 bool SbSystemSignWithCertificationSecretKey(const uint8_t* message,
                                             size_t message_size_in_bytes,
                                             uint8_t* digest,
                                             size_t digest_size_in_bytes) {
+  std::string secret_key_file_name = "/opt/cert_key";
+
+  ::starboard::ScopedFile file(secret_key_file_name.c_str(), O_RDONLY);
+  if ( !file.IsValid() ) {
+    SB_LOG(INFO) << "Cannot open secret key file '" << secret_key_file_name << "'";
+    return false;
+  }
+
+  auto sz = file.GetSize();
+  if ( (sz < 0) || (sz + 1 > 4096) ) {
+    SB_LOG(ERROR) << "Cannot read secret key contents of size: " << sz
+                  << " from: '" << secret_key_file_name << "'";
+    return false;
+  }
+
+  std::vector<char> buf;
+  buf.resize(sz + 1);
+  if ( file.ReadAll(buf.data(), sz) != sz ) {
+    SB_LOG(ERROR) << "Failed to read secret key contents of size: " << sz
+                  << " from: '" << secret_key_file_name << "'";
+    return false;
+  }
+  buf[sz] = 0;
+
+  return SignWithCertificationSecretKey(buf.data(), message, message_size_in_bytes, digest, digest_size_in_bytes);
+#if 0
   bool result = false;
 
 #if defined(HAS_WPECRYPTOGRAPHY)
@@ -174,4 +252,5 @@ bool SbSystemSignWithCertificationSecretKey(const uint8_t* message,
 #endif
 
   return result;
+#endif
 }
-- 
2.43.0

